File: app/src/main/java/hibernate/v2/ringtonerandomizer/ui/fragment/SelectRingtoneFragment.java
package hibernate.v2.ringtonerandomizer.ui.fragment;

import android.Manifest;
import android.app.Activity;
import android.content.DialogInterface;
import android.database.sqlite.SQLiteDatabase;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.afollestad.materialdialogs.MaterialDialog;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import hibernate.v2.ringtonerandomizer.C;
import hibernate.v2.ringtonerandomizer.R;
import hibernate.v2.ringtonerandomizer.helper.DBHelper;
import hibernate.v2.ringtonerandomizer.model.Ringtone;
import hibernate.v2.ringtonerandomizer.ui.adapter.RingtoneSelectAdapter;

public class SelectRingtoneFragment extends BaseFragment {

	protected final String[] PERMISSION_NAME = {
			Manifest.permission.READ_EXTERNAL_STORAGE
	};

	@BindView(R.id.rvlist)
	RecyclerView recyclerView;

	private ArrayList<Ringtone> allRingtoneList = new ArrayList<>();
	private ArrayList<Ringtone> shownRingtoneList = new ArrayList<>();
	private HashSet<String> hsFilter = new HashSet<>();
	private ArrayList<String> alFilter = new ArrayList<>();

	private String chosenPath = "";
	private MediaPlayer mediaPlayer;
	private boolean isPlaying = false;

	private DBHelper dbhelper;

	private SQLiteDatabase db;
	private RingtoneSelectAdapter ringtoneSelectAdapter;

	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
	                         Bundle savedInstanceState) {
		View rootView = inflater.inflate(R.layout.fragment_select_ringtone, container, false);
		ButterKnife.bind(this, rootView);
		return rootView;
	}

	@Override
	public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		openDatabase();

		recyclerView.setLayoutManager(new LinearLayoutManager(
				mContext, LinearLayoutManager.VERTICAL, false));

		ringtoneSelectAdapter = new RingtoneSelectAdapter(
				shownRingtoneList,
				this::openDialogPlayingPreview
		);
		recyclerView.setAdapter(ringtoneSelectAdapter);

		if (!isPermissionsGranted(PERMISSION_NAME)) {
			requestPermissions(PERMISSION_NAME, PERMISSION_REQUEST_CODE);
		}
	}

	@Override
	public void onResume() {
		super.onResume();

		if (isPermissionsGranted(PERMISSION_NAME)) {
			new LoadTask(this).execute();
		}
	}

	@Override
	public void onDestroy() {
		super.onDestroy();
		closeDatabase();
	}

	@Override
	public void onPause() {
		stopPlaying();
		super.onPause();
	}

	private void openDatabase() {
		dbhelper = new DBHelper(mContext);
		db = dbhelper.getWritableDatabase();
	}

	private void closeDatabase() {
		db.close();
		dbhelper.close();
	}

	private static class SaveTask extends AsyncTask<Void, Void, Void> {
		private MaterialDialog dialog;
		private WeakReference<Fragment> fragmentWeakReference;

		private SaveTask(Fragment fragment) {
			fragmentWeakReference = new WeakReference<>(fragment);
		}

		public void onPreExecute() {
			super.onPreExecute();

			SelectRingtoneFragment fragment = (SelectRingtoneFragment) fragmentWeakReference.get();
			if (fragment == null) {
				return;
			}

			Activity activity = fragment.getActivity();
			assert activity != null;

			dialog = new MaterialDialog.Builder(activity)
					.content(R.string.wait)
					.progress(true, 0)
					.cancelable(false)
					.show();
		}

		@Override
		public Void doInBackground(Void... arg0) {
			SelectRingtoneFragment fragment = (SelectRingtoneFragment) fragmentWeakReference.get();
			if (fragment == null) {
				return null;
			}

			for (Ringtone ringtone : fragment.ringtoneSelectAdapter.getCheckedRingtoneList()) {
				fragment.dbhelper.insertDBRingtone(ringtone);
			}
			return null;
		}

		public void onPostExecute(Void un) {
			SelectRingtoneFragment fragment = (SelectRingtoneFragment) fragmentWeakReference.get();
			if (fragment == null) {
				return;
			}

			Activity activity = fragment.getActivity();
			assert activity != null;

			dialog.dismiss();
			Toast.makeText(activity, R.string.done, Toast.LENGTH_SHORT).show();
			activity.finish();
		}
	}

	private static class LoadTask extends AsyncTask<Void, Void, Void> {
		private MaterialDialog dialog;
		private WeakReference<Fragment> fragmentWeakReference;

		private LoadTask(Fragment fragment) {
			fragmentWeakReference = new WeakReference<>(fragment);
		}

		public void onPreExecute() {
			super.onPreExecute();

			SelectRingtoneFragment fragment = (SelectRingtoneFragment) fragmentWeakReference.get();
			if (fragment == null) {
				return;
			}

			Activity activity = fragment.getActivity();
			assert activity != null;

			fragment.allRingtoneList.clear();
			fragment.shownRingtoneList.clear();

			dialog = new MaterialDialog.Builder(activity)
					.content(R.string.wait)
					.progress(true, 0)
					.cancelable(false)
					.show();
		}

		@Override
		public Void doInBackground(Void... arg0) {
			SelectRingtoneFragment fragment = (SelectRingtoneFragment) fragmentWeakReference.get();
			if (fragment == null) {
				return null;
			}

			Activity activity = fragment.getActivity();
			assert activity != null;

			ArrayList<Ringtone> ringtones = new ArrayList<>(C.getDeviceRingtoneList(activity));
			fragment.allRingtoneList = new ArrayList<>(ringtones);
			fragment.shownRingtoneList = new ArrayList<>(ringtones);
			return null;
		}

		public void onPostExecute(Void un) {
			SelectRingtoneFragment fragment = (SelectRingtoneFragment) fragmentWeakReference.get();
			if (fragment == null) {
				return;
			}

			fragment.ringtoneSelectAdapter.refreshData(fragment.shownRingtoneList);
			dialog.dismiss();
		}
	}

	private void filterList() {
		alFilter.clear();
		hsFilter.clear();
		hsFilter.add("Internal Storage");
		for (Ringtone ringtone : allRingtoneList) {
			int posStart = 1;
			int posEnd;
			while (true) {
				if (ringtone.getPath() == null) {
					break;
				}
				posEnd = ringtone.getPath().indexOf("/", posStart);
				if (posStart == posEnd || posEnd == -1)
					break;
				posStart = posEnd + 1;
				String stringSub = ringtone.getPath().substring(0, posEnd + 1);
				hsFilter.add(stringSub);
			}
		}
		alFilter.addAll(hsFilter);
		Collections.sort(alFilter);

		ArrayList<String> alFilterRingtone = new ArrayList<>(alFilter);
		ArrayList<String> alFilterTemp = new ArrayList<>(alFilter);
		for (String element : alFilter) {
			if (!element.toLowerCase().contains("ringtone")) {
				alFilterRingtone.remove(element);
			} else {
				alFilterTemp.remove(element);
			}
		}
		alFilter.clear();
		alFilter.addAll(alFilterRingtone);
		alFilter.addAll(alFilterTemp);
	}

	@OnClick(R.id.saveBtn)
	public void onClickSaveList() {
		new SaveTask(this).execute();
	}

	@OnClick(R.id.filter_button)
	public void onClickFilterDialog() {
		filterList();
		MaterialDialog.Builder dialog = new MaterialDialog.Builder(mContext)
				.title(R.string.filter_title)
				.items(alFilter)
				.itemsCallbackSingleChoice(-1, new MaterialDialog.ListCallbackSingleChoice() {
					@Override
					public boolean onSelection(MaterialDialog dialog, View view, int which, CharSequence text) {
						chosenPath = text.toString();
						shownRingtoneList.clear();
						for (Ringtone ringtone : allRingtoneList) {
							if (ringtone.getPath() == null) {
								if (chosenPath.equals("Internal")) {
									shownRingtoneList.add(ringtone);
								}
							} else {
								if (ringtone.getPath().contains(chosenPath)) {
									shownRingtoneList.add(ringtone);
								}
							}
						}
						ringtoneSelectAdapter.refreshData(shownRingtoneList);
						return true;
					}
				})
				.negativeText(R.string.item_navbtn);
		dialog.show();
	}

	private void openDialogPlayingPreview(Ringtone ringtone) {
		try {
			mediaPlayer = MediaPlayer.create(mContext, Uri.parse(ringtone.getUriId()));
			mediaPlayer.setLooping(true);
			mediaPlayer.start();

			String message;
			if (ringtone.getPath() != null) {
				message = getString(R.string.item_message) + ringtone.getName()
						+ "\n\n" + getString(R.string.item_message2) + ringtone.getPath();
			} else {
				message = getString(R.string.item_message) + ringtone.getName()
						+ "\n\n" + getString(R.string.item_message2) + "Internal Storage";
			}

			MaterialDialog.Builder dialog = new MaterialDialog.Builder(mContext)
					.title(R.string.playing_sound)
					.content(message)
					.dismissListener(new DialogInterface.OnDismissListener() {
						@Override
						public void onDismiss(DialogInterface dialog) {
							stopPlaying();
						}
					})
					.cancelable(false)
					.negativeText(R.string.item_navbtn);
			dialog.show();
		} catch (Exception e) {
			Toast.makeText(mContext, R.string.ui_error, Toast.LENGTH_SHORT).show();
		}
	}

	private void stopPlaying() {
		if (mediaPlayer != null) {
			if (isPlaying) {
				mediaPlayer.stop();
			}
			mediaPlayer.release();
			mediaPlayer = null;
			isPlaying = false;
		}
	}

	@Override
	public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
		super.onRequestPermissionsResult(requestCode, permissions, grantResults);
		if (requestCode == PERMISSION_REQUEST_CODE) {
			if (!hasAllPermissionsGranted(grantResults)) {
				C.openErrorPermissionDialog(mContext);
			}
		}
	}
}

File: app/src/main/java/hibernate/v2/ringtonerandomizer/ui/fragment/SelectRingtoneFragment.kt
package hibernate.v2.ringtonerandomizer.ui.fragment

import android.Manifest
import android.media.MediaPlayer
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.callbacks.onDismiss
import com.afollestad.materialdialogs.list.listItemsSingleChoice
import hibernate.v2.ringtonerandomizer.R
import hibernate.v2.ringtonerandomizer.helper.DBHelper
import hibernate.v2.ringtonerandomizer.helper.UtilHelper
import hibernate.v2.ringtonerandomizer.model.Ringtone
import hibernate.v2.ringtonerandomizer.model.Ringtone.Companion.PATH_INTERNAL_STORAGE
import hibernate.v2.ringtonerandomizer.ui.adapter.RingtoneSelectAdapter
import kotlinx.android.synthetic.main.fragment_select_ringtone.*
import java.lang.ref.WeakReference
import java.util.ArrayList

class SelectRingtoneFragment : BaseFragment() {

    private var allRingtoneList = ArrayList<Ringtone>()
    private var shownRingtoneList = ArrayList<Ringtone>()
    private val alFilter = ArrayList<String>()
    private var chosenPath = ""
    private var mediaPlayer: MediaPlayer? = null
    private var isPlaying = false
    private lateinit var ringtoneSelectAdapter: RingtoneSelectAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_select_ringtone, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        ringtoneSelectAdapter = RingtoneSelectAdapter(
                shownRingtoneList,
                object : RingtoneSelectAdapter.ItemClickListener {
                    override fun onItemDetailClick(ringtone: Ringtone) {
                        openDialogPlayingPreview(ringtone)
                    }
                })
        rvList.layoutManager = LinearLayoutManager(context)
        rvList.adapter = ringtoneSelectAdapter

        saveBtn.setOnClickListener { onClickSaveList() }
        filterBtn.setOnClickListener { onClickFilterDialog() }

        if (!isPermissionsGranted(PERMISSION_NAME)) {
            requestPermissions(PERMISSION_NAME, PERMISSION_REQUEST_CODE)
        }
    }

    override fun onResume() {
        super.onResume()
        if (isPermissionsGranted(PERMISSION_NAME)) {
            GetDeviceRingtoneTask(this).execute()
        }
    }

    override fun onPause() {
        stopPlaying()
        super.onPause()
    }

    private class SaveTask(fragment: SelectRingtoneFragment) : AsyncTask<Void?, Void?, Void?>() {
        private var dialog: MaterialDialog? = null
        private val fragmentWeakReference: WeakReference<SelectRingtoneFragment> = WeakReference(fragment)
        private var dbHelper = DBHelper(fragment.context)

        private val fragment: SelectRingtoneFragment?
            get() = fragmentWeakReference.get()

        public override fun onPreExecute() {
            super.onPreExecute()
            fragment?.activity?.let { activity ->
                dialog = MaterialDialog(activity)
                        .message(R.string.wait)
                        .cancelable(false)
                dialog?.show()
            }
        }

        override fun doInBackground(vararg param: Void?): Void? {
            fragment?.ringtoneSelectAdapter?.checkedRingtoneList?.let { checkedRingtoneList ->
                for (ringtone in checkedRingtoneList) {
                    dbHelper.insertDBRingtone(ringtone)
                }
            }

            return null
        }

        public override fun onPostExecute(un: Void?) {
            dbHelper.close()
            Toast.makeText(fragment?.activity, R.string.done, Toast.LENGTH_SHORT).show()
            fragment?.activity?.let { activity ->
                if (activity.isDestroyed) return
                dialog?.dismiss()
                activity.finish()
            }
        }
    }

    private class GetDeviceRingtoneTask(fragment: SelectRingtoneFragment) : AsyncTask<Void?, Void?, Void?>() {
        private var dialog: MaterialDialog? = null
        private val fragmentWeakReference: WeakReference<SelectRingtoneFragment> = WeakReference(fragment)

        private val fragment: SelectRingtoneFragment?
            get() = fragmentWeakReference.get()

        public override fun onPreExecute() {
            super.onPreExecute()
            fragment?.allRingtoneList?.clear()
            fragment?.shownRingtoneList?.clear()
            fragment?.activity?.let { activity ->
                dialog = MaterialDialog(activity)
                        .message(R.string.wait)
                        .cancelable(false)
                dialog?.show()
            }
        }

        override fun doInBackground(vararg param: Void?): Void? {
            val ringtones = UtilHelper.getDeviceRingtoneList(fragment?.activity)
            // Clone the ArrayList
            fragment?.allRingtoneList = ArrayList(ringtones)
            fragment?.shownRingtoneList = ArrayList(ringtones)
            return null
        }

        public override fun onPostExecute(un: Void?) {
            fragment?.activity?.let { activity ->
                fragment?.shownRingtoneList?.let { shownRingtoneList ->
                    fragment?.ringtoneSelectAdapter?.refreshData(shownRingtoneList)
                }
                if (activity.isDestroyed) return
                dialog?.dismiss()
            }
        }
    }

    private fun filterList() {
        alFilter.clear()
        val hsFilter = hashSetOf(PATH_INTERNAL_STORAGE)
        for (ringtone in allRingtoneList) {
            var posStart = 1
            var posEnd: Int
            ringtone.path?.let {
                while (true) {
                    posEnd = it.indexOf("/", posStart)
                    if (posStart == posEnd || posEnd == -1) break
                    posStart = posEnd + 1
                    val stringSub = it.substring(0, posEnd + 1)
                    hsFilter.add(stringSub)
                }
            }
        }
        alFilter.addAll(hsFilter)
        alFilter.sort()
        val alFilterRingtone = ArrayList(alFilter)
        val alFilterTemp = ArrayList(alFilter)
        for (element in alFilter) {
            if (!element.toLowerCase().contains("ringtone")) {
                alFilterRingtone.remove(element)
            } else {
                alFilterTemp.remove(element)
            }
        }
        alFilter.clear()
        alFilter.addAll(alFilterRingtone)
        alFilter.addAll(alFilterTemp)
    }

    private fun onClickSaveList() {
        SaveTask(this).execute()
    }

    private fun onClickFilterDialog() {
        filterList()
        context?.let { context ->
            MaterialDialog(context)
                    .title(R.string.filter_title)
                    .listItemsSingleChoice(items = alFilter, waitForPositiveButton = false) { dialog, _, text ->
                        chosenPath = text.toString()
                        shownRingtoneList.clear()
                        for (ringtone in allRingtoneList) {
                            ringtone.path?.let { path ->
                                if (path.contains(chosenPath)) {
                                    shownRingtoneList.add(ringtone)
                                }
                            } ?: run {
                                if (chosenPath == PATH_INTERNAL_STORAGE) {
                                    shownRingtoneList.add(ringtone)
                                }
                            }
                        }
                        ringtoneSelectAdapter.refreshData(shownRingtoneList)
                        currentPathTv.text = chosenPath
                        dialog.dismiss()
                    }
                    .neutralButton(R.string.ui_reset) { dialog ->
                        ringtoneSelectAdapter.refreshData(allRingtoneList)
                        currentPathTv.text = "N/A"
                        dialog.dismiss()
                    }
                    .negativeButton(R.string.item_navbtn)
                    .show()
        }
    }

    private fun openDialogPlayingPreview(ringtone: Ringtone) {
        try {
            mediaPlayer = MediaPlayer.create(context, Uri.parse(ringtone.uriId))
            mediaPlayer?.isLooping = true
            mediaPlayer?.start()
            val message: String = if (ringtone.path != null) {
                (getString(R.string.item_message) + ringtone.name
                        + "\n\n" + getString(R.string.item_message2) + ringtone.path)
            } else {
                (getString(R.string.item_message) + ringtone.name
                        + "\n\n" + getString(R.string.item_message2) + PATH_INTERNAL_STORAGE)
            }
            context?.let { context ->
                MaterialDialog(context)
                        .title(R.string.playing_sound)
                        .message(text = message)
                        .onDismiss { stopPlaying() }
                        .negativeButton(R.string.item_navbtn)
                        .show()
            }
        } catch (e: Exception) {
            Toast.makeText(context, R.string.ui_error, Toast.LENGTH_SHORT).show()
        }
    }

    private fun stopPlaying() {
        if (isPlaying) {
            mediaPlayer?.stop()
        }
        mediaPlayer?.release()
        mediaPlayer = null
        isPlaying = false
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == PERMISSION_REQUEST_CODE) {
            if (!hasAllPermissionsGranted(grantResults)) {
                UtilHelper.openErrorPermissionDialog(context)
            }
        }
    }

    companion object {
        private val PERMISSION_NAME = arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)
    }
}